p8e-util/src/main/kotlin/io/p8e/crypto/Pen.kt

```diff
@@ -64,19 +64,61 @@ class Pen(
             .orThrow { IllegalStateException("can't verify signature - public cert may not match private key.") }
     }
 
-    override fun sign(): ByteArray = signature.sign()
-
-    override fun update(data: ByteArray) { signature.update(data) }
+    override fun sign(): ByteArray {
+        signature.update(aggregatedData)
+        return signature.sign()
+    }
 
-    override fun update(data: ByteArray, off: Int, len: Int) { signature.update(data, off, len) }
+    override fun update(data: ByteArray) = signature.update(data)
+
+    override fun update(data: ByteArray, off: Int, res: Int) {
+        // If off is less then res, these are the data that we care about.
+        if(off < res) {
+            if(!verifying) {
+                val dataSample = data.copyOfRange(off, off+res)
```

Comment:
This might help reduce the size of the objects we upload to OS, when I was sampling the data with the legacy way of signing (Signature.getInstace() ... ) this included a bunch of trailing insignificant bytes as though it was just empty space. The `off` and the `res` here tells us where the data that we need starts and ends.

Reply:
What does `res` stand for? rest? residual? restraint? resign? resplendent? respiratory? resolute?

Reply:
I don't know, maybe `result`, just took what was set here: https://github.com/provenance-io/p8e/blob/8ac68c7363d4bcec02c6c147f55e85dc6bc6099f/os-client/src/main/kotlin/io/provenance/os/domain/inputstream/SignatureInputStream.kt#L33 - so i don't confuse the value with `len`

Reply:
oh, must just be the length of bytes actually read, k